[Merged by Bors] - feat(NumberTheory): every number field has a ramified prime

erdOne opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

erdOne committed:
feat(NumberTheory): every number field has a ramified prime

erdOne committed:
fix

erdOne committed:
fix

erdOne committed:
fix

erdOne committed:
fix

xroblot commented on Mathlib/NumberTheory/NumberField/Discriminant/Different.lean:
```diff
@@ -20,12 +20,14 @@ import Mathlib.Tactic.Qify
 open nonZeroDivisors
 
-lemma NumberField.absNorm_differentIdeal : (differentIdeal ℤ 𝒪).absNorm = (discr K).natAbs := by
+lemma NumberField.absNorm_differentIdeal [Module.Free ℤ 𝒪] [IsDedekindDomain 𝒪] :
```
Don't you think it would be convenient to make `K` explicit here? To tell you the truth, I used this result in some future PRs and it is one of the first thing I did ;)

xroblot commented on Mathlib/NumberTheory/NumberField/Discriminant/Different.lean:
```diff
@@ -61,11 +63,45 @@ lemma NumberField.absNorm_differentIdeal : (differentIdeal ℤ 𝒪).absNorm = (
+
+lemma NumberField.not_dvd_discr_iff_forall_liesOver {p : ℤ} (hp : Prime p) :
+    ¬ p ∣ discr K ↔ ∀ (P : Ideal 𝒪) (_ : P.IsMaximal), P.LiesOver (.span {p}) →
+      Algebra.IsUnramifiedAt ℤ P := by
```
Maybe we should introduce something like `Algebra.IsUnramifiedOver` that says what this long line of implication is saying

riccardobrasca replied:
This seems a good idea. Is it possible to have it in this PR?

xroblot commented on Mathlib/NumberTheory/RamificationInertia/Galois.lean:
```diff
@@ -148,7 +148,7 @@ theorem ramificationIdx_eq_of_isGalois [IsGalois K L] :
 
 /-- All the `inertiaDeg` over a fixed maximal ideal are the same. -/
-theorem inertiaDeg_eq_of_isGalois [p.IsMaximal] [IsGalois K L] :
+theorem inertiaDeg_eq_of_isGalois [IsGalois K L] :
```
You should change the docstring too (BTW, it's great that you could remove the `IsMaximal` hypothesis)

xroblot commented on Mathlib/RingTheory/Ideal/Norm/AbsNorm.lean:
```diff
@@ -343,6 +344,58 @@ theorem absNorm_ne_zero_of_nonZeroDivisors (I : (Ideal S)⁰) : absNorm (I : Ide
+See `Ideal.absNorm_pow_inertiaDeg`. -/
+lemma exists_prime_and_absNorm_eq_pow (P : Ideal S) [P.IsMaximal] :
+    ∃ p n, 0 < n ∧ ↑p ∈ P ∧ p.Prime ∧ P.absNorm = p ^ n := by
+  have hP : P ≠ ⊥ := by
```
Could you make a very useful lemma from this `have`?

xroblot reviewed (approved):
LGTM. I only have some minor remarks

xroblot commented on Mathlib/NumberTheory/RamificationInertia/Basic.lean:
```diff
@@ -307,29 +311,43 @@ lemma inertiaDeg_comap_eq (e : S ≃ₐ[R] S₁) (P : Ideal S₁) [p.IsMaximal]
+    [IsDedekindDomain R] [Module.Free ℤ R] [Module.Finite ℤ R]
+    [IsDedekindDomain S] [Module.Free ℤ S] [Module.Finite ℤ S]
+    [Algebra R S] [p.IsPrime] [P.LiesOver p] :
+    p.absNorm ^ p.inertiaDeg P = P.absNorm := by
```
I do not have strong feelings either. It's just that the way it is stated in `absNorm_eq_pow_inertiaDeg` is the one I found most useful in my applications. Anyway, both lemmas should be a the same direction. As it is your PR, I let decide which direction 😛

erdOne replied:
I swapped the two `pow_inertiaDeg_eq_absNorm` lemmas. Most invocations are actually rewriting in the inverse direction.

erdOne commented on Mathlib/RingTheory/Ideal/Norm/AbsNorm.lean:
```diff
@@ -343,6 +344,58 @@ theorem absNorm_ne_zero_of_nonZeroDivisors (I : (Ideal S)⁰) : absNorm (I : Ide
+  letI := Ideal.finiteQuotientOfFreeOfNeBot P hP
+  cases nonempty_fintype (S ⧸ P)
+  letI := Ideal.Quotient.field P
+  obtain ⟨p, hpR⟩ := CharP.exists (S ⧸ P)
```
Yeah I couldn't find a better proof without more import so I just put this here and point people to `Ideal.absNorm_pow_inertiaDeg` in the docstring.

erdOne commented on Mathlib/RingTheory/Ideal/Norm/AbsNorm.lean:
```diff
@@ -343,6 +344,58 @@ theorem absNorm_ne_zero_of_nonZeroDivisors (I : (Ideal S)⁰) : absNorm (I : Ide
+  exact P.absNorm_mem
+
+lemma exists_isMaximal_dvd_of_dvd_absNorm
+    {p : ℤ} (hp : Prime p) (I : Ideal S) (hI : p ∣ I.absNorm) :
```
I think this is more useful (and easier to prove) as `p : ℤ` though. I can provide a `p : ℕ` version as an easy corollary if you prefer.

erdOne committed:
address comments

erdOne committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into erd1/existsRamified

erdOne committed:
sorry new file

erdOne committed:
maybe not

erdOne committed:
oops

erdOne committed:
fix

erdOne committed:
add variant

erdOne committed:
address comments

erdOne committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into erd1/existsRamified

erdOne committed:
fix

erdOne committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into erd1/existsRamified

jcommelin commented on Mathlib/NumberTheory/NumberField/ExistsRamified.lean:
```diff
@@ -0,0 +1,59 @@
+
+/-!
+# Every number field has a ramified prime over `ℚ`
+...except `ℚ` itself.
```
I'm wondering... is the contrapositive useful as well?

A number field that is unramified over Q has finrank = 1.

erdOne committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into erd1/existsRamified

erdOne committed:
fix

erdOne committed:
oops

erdOne committed:
fix

erdOne committed:
fix

erdOne committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into erd1/existsRamified

erdOne committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into erd1/existsRamified

erdOne committed:
Merge branch 'erd1/existsRamified' of https://github.com/erdOne/mathlib4 into erd1/existsRamified

erdOne committed:
revert

erdOne committed:
more revert

erdOne committed:
fix

erdOne committed:
fix?

erdOne committed:
stuff

erdOne committed:
and

erdOne committed:
fix

erdOne committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into erd1/existsRamified

erdOne committed:
fix

erdOne committed:
fix

erdOne committed:
fix?

erdOne committed:
address comments

erdOne committed:
move

erdOne committed:
revert

erdOne committed:
fix

riccardobrasca commented:
@erdOne are you still interested in this? I think we really want the connection between ramification and the discriminant.

erdOne committed:
fix

erdOne committed:
Met push
rge branch 'master' of https://github.com/leanprover-community/mathlib4 into erd1/existsRamified

erdOne commented:
I would love any help on getting this merged. Feel free to push directly to this branch etc.

erdOne committed:
fix

tb65536 commented on Mathlib/LinearAlgebra/Dimension/FreeAndStrongRankCondition.lean:
```diff
@@ -295,4 +295,14 @@ alias ⟨_, bot_eq_top_of_finrank_eq_one⟩ := bot_eq_top_iff_finrank_eq_one
+    exact fun h ↦ not_nontrivial E h.injective.nontrivial
+  rw [← Subalgebra.bot_eq_top_iff_finrank_eq_one, Function.Bijective,
+    Algebra.surjective_algebraMap_iff, eq_comm]
+  simp [FaithfulSMul.algebraMap_injective]
```
```suggestion
  refine ⟨?_, Module.finrank_of_bijective_algebraMap⟩
  nontriviality E
  refine fun h ↦ ⟨FaithfulSMul.algebraMap_injective F E, ?_⟩
  rwa [Algebra.surjective_algebraMap_iff, eq_comm, Subalgebra.bot_eq_top_iff_finrank_eq_one]
```

erdOne committed:
Update Mathlib/LinearAlgebra/Dimension/FreeAndStrongRankCondition.lean

Co-authored-by: Thomas Browning <tb65536@users.noreply.github.com>

tb65536 commented on Mathlib/NumberTheory/NumberField/Discriminant/Different.lean:
```diff
@@ -166,4 +169,39 @@ theorem natAbs_discr_eq_natAbs_discr_pow_mul_natAbs_discr_pow (K₁ K₂ : Inter
+  · intro H h
+    rw [← Int.dvd_natAbs, ← absNorm_differentIdeal K 𝒪] at h
+    obtain ⟨P, hP, h₁, h₂⟩ := Ideal.exists_isMaximal_dvd_of_dvd_absNorm hp _ h
+    exact H P hP ⟨h₁.symm⟩ h₂
```
Minor, but this makes the logic slightly clearer:
```suggestion
  contrapose!
  constructor
  · intro h
    rw [← Int.dvd_natAbs, ← absNorm_differentIdeal K 𝒪] at h
    obtain ⟨P, hP, h₁, h₂⟩ := Ideal.exists_isMaximal_dvd_of_dvd_absNorm hp _ h
    exact ⟨P, hP, ⟨h₁.symm⟩, h₂⟩
  · rintro ⟨P, hP, hP', hP''⟩
    have := Ideal.absNorm_dvd_absNorm_of_le (Ideal.dvd_iff_le.mp hP'')
    rw [absNorm_differentIdeal K, Ideal.absNorm_eq_pow_inertiaDeg P hp,
      ← Int.natAbs_pow, Int.natAbs_dvd_natAbs] at this
    exact (dvd_pow_self _ (Ideal.inertiaDeg_pos' ..).ne').trans this
```

erdOne committed:
Update Mathlib/NumberTheory/NumberField/Discriminant/Different.lean

Co-authored-by: Thomas Browning <tb65536@users.noreply.github.com>

tb65536 commented on Mathlib/RingTheory/Ideal/Norm/AbsNorm.lean:
```diff
@@ -345,6 +346,63 @@ theorem absNorm_ne_zero_of_nonZeroDivisors (I : (Ideal S)⁰) : absNorm (I : Ide
+    cases hI with
+    | inr h =>
+      obtain ⟨Q, h₁, h₂, h₃⟩ := IH h
+      refine ⟨Q, h₁, h₂, dvd_mul_of_dvd_right h₃ _⟩
```
```suggestion
      exact ⟨Q, h₁, h₂, dvd_mul_of_dvd_right h₃ _⟩
```

erdOne committed:
Update Mathlib/RingTheory/Ideal/Norm/AbsNorm.lean

Co-authored-by: Thomas Browning <tb65536@users.noreply.github.com>

tb65536 commented on Mathlib/RingTheory/IntegralClosure/IsIntegralClosure/Basic.lean:
```diff
@@ -630,6 +631,11 @@ theorem Algebra.IsIntegral.isField_iff_isField [IsDomain S]
     (hRS : Function.Injective (algebraMap R S)) : IsField R ↔ IsField S :=
   ⟨isField_of_isIntegral_of_isField', isField_of_isIntegral_of_isField hRS⟩
 
+theorem Ideal.IsMaximal.ne_bot_of_isIntegral_int {R : Type*} [CommRing R]
```
`R` already exists in the context (twice!)
```suggestion
theorem Ideal.IsMaximal.ne_bot_of_isIntegral_int
```

tb65536 reviewed (approved):
maintainer delegate

erdOne committed:
Update Mathlib/RingTheory/IntegralClosure/IsIntegralClosure/Basic.lean

Co-authored-by: Thomas Browning <tb65536@users.noreply.github.com>

MichaelStollBayreuth commented on Mathlib/NumberTheory/NumberField/ExistsRamified.lean:
```diff
@@ -0,0 +1,89 @@
+  exact bijective_algebraMap_of_linearEquiv (IsIntegralClosure.equiv ℤ ℤ K 𝒪).toLinearEquiv
+
+/-- If `K` is a number field with positive rank such that `K/ℚ` is galois, then there exists
+some rational prime `p : ℕ` such that every prime of `K` over `P` is unramified. -/
```
```suggestion
some rational prime `p : ℕ` such that every prime of `K` over `P` is ramified. -/
```
I guess?

erdOne committed:
Update Mathlib/NumberTheory/NumberField/ExistsRamified.lean

Co-authored-by: Michael Stoll <99838730+MichaelStollBayreuth@users.noreply.github.com>

xroblot commented on Mathlib/NumberTheory/NumberField/ExistsRamified.lean:
```diff
@@ -0,0 +1,89 @@
+
+/-- If `K` is a number field with positive rank, then there exists some maximal ideal of `𝓞 K`
+that is ramified over `ℤ`. -/
+lemma NumberField.exists_not_isUramifiedAt_int (H : Module.finrank ℚ K ≠ 1) :
```
```suggestion
lemma NumberField.exists_not_isUnramifiedAt_int (H : Module.finrank ℚ K ≠ 1) :
```

xroblot commented on Mathlib/NumberTheory/NumberField/ExistsRamified.lean:
```diff
@@ -0,0 +1,89 @@
+
+/-- If `K` is a number field with positive rank such that `K/ℚ` is galois, then there exists
+some rational prime `p : ℕ` such that every prime of `K` over `P` is ramified. -/
+lemma NumberField.exists_not_isUramifiedAt_int_of_isGalois [IsGalois ℚ K]
```
```suggestion
lemma NumberField.exists_not_isUnramifiedAt_int_of_isGalois [IsGalois ℚ K]
```

xroblot commented on Mathlib/NumberTheory/NumberField/ExistsRamified.lean:
```diff
@@ -0,0 +1,89 @@
+# Every number field has a ramified prime over `ℚ`
+...except `ℚ` itself.
+
+This is a trivial corollary of `NumberField.not_dvd_discr_iff_forall_pow_mem` and
```
```suggestion
This is a trivial corollary of `NumberField.not_dvd_discr_iff_forall_mem` and
```

xroblot commented on Mathlib/NumberTheory/NumberField/ExistsRamified.lean:
```diff
@@ -0,0 +1,89 @@
+  exact bijective_algebraMap_of_linearEquiv (IsIntegralClosure.equiv ℤ ℤ K 𝒪).toLinearEquiv
+
+/-- If `K` is a number field with positive rank such that `K/ℚ` is galois, then there exists
+some rational prime `p : ℕ` such that every prime of `K` over `P` is ramified. -/
```
```suggestion
some rational prime `p : ℕ` such that every prime of `K` over `p` is ramified. -/
```

xroblot commented:
LGTM, although I still think having something like an abbrev `Algebra.IsUnramifiedOver` as mentioned above would go a long way to making the statement look nicer, but that can wait for another PR.

riccardobrasca commented:
Thanks!

bors d+

erdOne committed:
Apply suggestions from code review

Co-authored-by: Xavier Roblot <46200072+xroblot@users.noreply.github.com>

erdOne committed:
Merge branch 'master' into erd1/existsRamified

erdOne committed:
fix

erdOne committed:
Update Mathlib/NumberTheory/NumberField/ExistsRamified.lean

Co-authored-by: Xavier Roblot <46200072+xroblot@users.noreply.github.com>

erdOne committed:
oops
